overlord/devicestate/handlers_install.go

```diff
@@ -618,3 +631,255 @@ func (m *DeviceManager) doRestartSystemToRunMode(t *state.Task, _ *tomb.Tomb) er
 
 	return nil
 }
+
+func (m *DeviceManager) doFactoryResetRunSystem(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	perfTimings := state.TimingsForTask(t)
+	defer perfTimings.Save(st)
+	// get gadget dir
+	deviceCtx, err := DeviceCtx(st, t, nil)
+	if err != nil {
+		return fmt.Errorf("cannot get device context: %v", err)
+	}
+	gadgetInfo, err := snapstate.GadgetInfo(st, deviceCtx)
+	if err != nil {
+		return fmt.Errorf("cannot get gadget info: %v", err)
+	}
+	gadgetDir := gadgetInfo.MountDir()
+
+	kernelInfo, err := snapstate.KernelInfo(st, deviceCtx)
+	if err != nil {
+		return fmt.Errorf("cannot get kernel info: %v", err)
+	}
+	kernelDir := kernelInfo.MountDir()
+
+	modeEnv, err := maybeReadModeenv()
+	if err != nil {
+		return err
+	}
+	if modeEnv == nil {
+		return fmt.Errorf("missing modeenv, cannot proceed")
+	}
+
+	// bootstrap
+	bopts := install.Options{
+		Mount: true,
+	}
+	encryptionType, err := m.checkEncryption(st, deviceCtx)
+	if err != nil {
+		return err
+	}
+	bopts.EncryptionType = encryptionType
+	useEncryption := (encryptionType != secboot.EncryptionTypeNone)
+	hasMaker := osutil.FileExists(filepath.Join(boot.InstallHostFDESaveDir, "marker"))
+	// TODO verify that the same encryption mechanism is used
+	if hasMaker != useEncryption {
+		prevStatus := "encrypted"
+		if !hasMaker {
+			prevStatus = "unencrypted"
+		}
+		// the original system was (un)encrypted
+		return fmt.Errorf("cannot perform factory reset using different encryption, the original system was %v", prevStatus)
```

Comment:
I think that the error message is more clear than the comment, which I'd rather remove.